### PR TITLE
Fixes #590, #605 Custom Filtering

### LINKED ISSUE #590: Implement extended filter functionality for ElementDefinition browser
### Prerequisites

- [x] I have written a descriptive issue title
- [x] I have verified that I am running the latest version of the CDP4-IME
- [x] I have searched [open](https://github.com/RHEAGROUP/CDP4-IME-Community-Edition/issues) and [closed](https://github.com/RHEAGROUP/CDP4-IME-Community-Edition/issues?q=is%3Aissue+is%3Aclosed) issues to ensure it has not already been reported

### Description
Implement extended filter functionality in ElementDefinitionBrowser:

- ClassKind
- Owner
- Applied Categories
- ParameterType
- MemberOf

Ensure that full hierarchy is displayed on filtered items.

### PR DESCRIPTION
**Pull request by lxatrhea**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/RHEAGROUP/CDP4-IME-Community-Edition/pulls) open
- [x] I have verified that I am following the CDP4-IME [code style guidelines](https://raw.githubusercontent.com/RHEAGROUP/CDP4-IME-Community-Edition/master/.github/CONTRIBUTING.md)
- [x] I have provided test coverage for my change (where applicable)

### Description
Add Custom Filter functionality to all Browsers.

### CODE REVIEW

File: CDP4Composition/Services/FilterEditorService/ISavedUserPreference.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+//    GNU Affero General Public License for more details.
+//
+//    You should have received a copy of the GNU Affero General Public License
+//    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+// </copyright>
+// -------------------------------------------------------------------------------------------------
+
+namespace CDP4Composition.Services.FilterEditorService
+{
+    public interface ISavedUserPreference
```
Comment (alexatrhea) [inline CDP4Composition/Services/FilterEditorService/ISavedUserPreference.cs:28]: docs

File: EngineeringModel/ViewModels/ElementDefinitionBrowser/ElementDefinitionTreeRows/ElementUsageRowViewModel.cs [hunk 2]
```diff
@@ -147,7 +147,9 @@ public ReactiveList<Option> SelectedOptions
             get { return this.selectedOptions; }
             set { this.RaiseAndSetIfChanged(ref this.selectedOptions, value); }
         }
-        
+
+        public new IEnumerable<Category> Category => this.Thing.Category.Union(this.Thing.ElementDefinition.Category).Distinct();
```
Comment (alexatrhea) [inline EngineeringModel/ViewModels/ElementDefinitionBrowser/ElementDefinitionTreeRows/ElementUsageRowViewModel.cs:151]: docs

File: EngineeringModel/ViewModels/ElementDefinitionBrowser/ElementDefinitionTreeRows/IElementBaseRowViewModel.cs [hunk 3]
```diff
@@ -1,18 +1,23 @@
...
+
     using CDP4Common.EngineeringModelData;
+    using CDP4Common.SiteDirectoryData;
 
     /// <summary>
     /// The interface for rows that contains rows representing a <see cref="ParameterGroup"/> or <see cref="ParameterBase"/>
     /// </summary>
     public interface IElementBaseRowViewModel : IModelCodeRowViewModel
     {
+        IEnumerable<Category> Category { get; }
```
Comment (alexatrhea) [inline EngineeringModel/ViewModels/ElementDefinitionBrowser/ElementDefinitionTreeRows/IElementBaseRowViewModel.cs:19]: docs

**Review by alexatrhea — CHANGES_REQUESTED** [review]
Docs missing here and there, for the rest, good.

**Review by alexatrhea — APPROVED** [review]

**Review by alexatrhea — APPROVED** [review]

File: EngineeringModel/ViewModels/ElementDefinitionBrowser/ElementDefinitionTreeRows/ElementBaseRowViewModel.cs [hunk 4]
```diff
@@ -147,6 +147,8 @@ public string ModelCode
             private set { this.RaiseAndSetIfChanged(ref this.modelCode, value); }
         }
 
+        public IEnumerable<Category> Category => this.Thing.Category;
```
Comment (samatrhea) [inline EngineeringModel/ViewModels/ElementDefinitionBrowser/ElementDefinitionTreeRows/ElementBaseRowViewModel.cs:150]: docs

**Review by samatrhea — APPROVED** [review]
some small comments

**Pull request merged**